lib/mongo/server/connection.rb

```diff
@@ -155,7 +155,15 @@ def ping
 
       def deliver(messages)
         write(messages)
-        messages.last.replyable? ? read : nil
+
+        if messages.last.replyable?
+          # Protection against returning the response to a previous request. See
+          # RUBY-1117
+          reply = read until reply && reply.response_to == messages.last.request_id
```

Comment:
This is actually very close to our first attempt at fixing the issue.
The problem with that approach is that we would need to be absolutely certain the socket will also be closed when we read the wrong response from the network.

If not, then we just push back the problem to the next read, indefinitely, and that would not solve anything.

Also, it would be nice to give the process a chance to recover before causing it to fail. If not then it just pushes back the error handling responsibility on the app instead of the library. Maybe we could try reading just once more and then abort? It would be stupid to drop the socket and it's pending messages if our reply is in there.

IMO the best behavior would be to read until we get our response, and once there is nothing more to read, the request should timeout. That is actually linked to https://jira.mongodb.org/browse/RUBY-1120 IMO, which is the opposite problem, the response from mongodb is missed or lost and the client waits indefinitely.

We had a look at the code with @elhu and saw there is a socket timeout option which defaults at 5s and is set as read and write timeout on the socket using `setsockopt`, unfortunately this seems to be broken in ruby itself, because when you do `io.read(16)` in ruby, even if the `read` syscall times out (we see it returning -1 in `strace`), ruby continue waiting (probably with a `select` or `ppoll` as seen in RUBY-1120) which means this timeout has no impact.

If this timeout were working properly, it wouldn't be a problem to read multiple times.
Of course if you really prefer this way it's better than nothing, but it means everyone doing thread kill or on bad network will get random exceptions in production and broken builds and will have no way to improve this.

ps: about passing the `request_id` argument, to me it doesn't sounds like it's the `deserialize` method responsibility to check this, it is supposed to just get from wire protocol to `Message` object. Same for the `read` method. So even if this should raise an exception, It looks more appropriate in the deliver method to me (with the `response_to` added to the message as it is part but was just ignored). But that's just my opinion.